library/src/cache/patch_manager.rs

```diff
@@ -0,0 +1,883 @@
+
+        std::fs::create_dir_all(self.patch_dir(patch_number))
+            .with_context(|| format!("create_dir_all failed for {}", patch_path.display()))?;
+
+        std::fs::rename(file_path, &patch_path)?;
+
+        let new_patch = PatchMetadata {
+            number: patch_number,
+            size: std::fs::metadata(&patch_path)?.len(),
+        };
+
+        // If a patch was never booted (next_boot_patch != last_booted_patch), we should delete
+        // it here before setting next_boot_patch to the new patch.
+        if let (Some(last_boot_patch), Some(next_boot_patch)) = (
+            self.patches_state.next_boot_patch,
+            self.patches_state.last_booted_patch,
+        ) {
+            if last_boot_patch.number != next_boot_patch.number {
+                if let Err(e) = self.delete_patch_artifacts(next_boot_patch.number) {
+                    error!(
+                        "Failed to delete patch artifacts for patch {}. Error: {}",
+                        patch_number, e
+                    );
+                }
+            }
+        }
+
+        self.patches_state.next_boot_patch = Some(new_patch);
+        self.patches_state.highest_seen_patch_number = self
+            .patches_state
+            .highest_seen_patch_number
+            .map(|highest_patch_number: usize| highest_patch_number.max(patch_number))
+            .or(Some(patch_number));
+        self.save_patches_state()
+    }
+
+    fn last_successfully_booted_patch(&self) -> Option<PatchInfo> {
+        self.patches_state
+            .last_booted_patch
+            .map(|patch| self.patch_info_for_number(patch.number))
+    }
+
+    fn next_boot_patch(&mut self) -> Option<PatchInfo> {
+        let next_boot_patch = match self.patches_state.next_boot_patch {
+            Some(patch) => patch,
+            None => return None,
+        };
+
+        if let Err(e) = self.validate_patch_is_bootable(&next_boot_patch) {
+            error!("Patch {} is not bootable: {}", next_boot_patch.number, e);
+
+            self.try_fall_back_to_last_booted_patch();
+
+            if let Err(e) = self.save_patches_state() {
+                error!("Failed to save patches state: {}", e);
+            }
+            return None;
+        }
+
+        self.patches_state
+            .next_boot_patch
+            .as_ref()
+            .map(|patch| self.patch_info_for_number(patch.number))
+    }
+
+    fn record_boot_success_for_patch(&mut self, patch_number: usize) -> Result<()> {
+        let next_boot_patch = self
+            .patches_state
+            .next_boot_patch
+            .context("No next_boot_patch")?;
+
+        if next_boot_patch.number != patch_number {
+            bail!(
+                "Attempted to record boot success for patch {} but next_boot_patch is {}",
+                patch_number,
+                next_boot_patch.number
+            );
+        }
+
+        if let Some(current_patch) = self.patches_state.last_booted_patch {
```

Comment:
So a side-effect of successfully booting is that we clean up previously booted patches?

Reply:
Couldn't this end up cleaning up the "next_boot_patch" if they're the same?

Reply:
Yeah, assuming I'm reading this correctly, this method is such that we only get to boot once from a patch and then it deletes the patch 🤣   Am I reading that wrong?

Reply:
Nope, you're reading correctly. We just happened to only call this function for patches that had not previously booted (probably incorrectly). Fixed!